[STEP2-4] Add Feature Blocking domains

sunsun512 opened the pull request:
## 변동사항

> 특정 도메인은 shorten-url 제공을 거부하고 싶어요.

-  csv 파일을 API 통해 업로드 받아서, 해당 파일에 있는 도메인 목록을 서비스 미제공 도메인으로 등록합니다.

sunsun512 committed:
feat: block domain 로직 추가

sunsun512 committed:
feat: block domain 로직 추가

sunsun512 committed:
feat: block domain 로직 추가

sunsun512 committed:
feat: .gitignore 수정

sunsun512 committed:
Delete src/main/resources/application.yml

sunsun512 committed:
Delete src/main/resources/application-dev.yml

Hyune-c commented on src/main/java/community/whatever/onembackendjava/api/service/UrlShortenService.java:
```diff
@@ -1,21 +1,24 @@
+@RequiredArgsConstructor
 public class UrlShortenService {
 
+    private final BlackListService blackListService;
```
service 간 호출이 이루어지고 있네요. service 레이어의 구분에 대해 고민해주세요.
- fyi. https://msolo021015.medium.com/layered-architecture-deep-dive-c0a5f5a9aa37

sunsun512 replied:
BlockedDomainInterface 으로 레이어를 구분했습니다.

Hyune-c commented on src/main/java/community/whatever/onembackendjava/api/service/UrlShortenService.java:
```diff
@@ -54,6 +60,15 @@ public static boolean isValidURL(String url) {
         return matcher.matches();
     }
 
+    public boolean isBlocked(String url) {
+        for (String domain : blackListService.getBlockedDomains()) {
+            if (url.contains(domain)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
```
```suggestion
    public boolean isBlocked(String url) {
        return blackListService.getBlockedDomains().parallelStream()
                .anyMatch(url::contains);
    }
```

`nit`

Hyune-c commented on src/main/java/community/whatever/onembackendjava/api/service/UrlShortenService.java:
```diff
@@ -35,6 +38,9 @@ public ShortenUrlDto.Create.Response shortenUrlCreate(ShortenUrlDto.Create.Reque
         }
+        if(isBlocked(param.getOriginUrl())){
+            throw BusinessExceptionGenerator.createBusinessException("DB004");
+        }
```
이런 형식이라면 ErrorCode enum 을 매개변수로 받는게 낫지 않나요?

Hyune-c commented on src/main/java/community/whatever/onembackendjava/common/util/BlackListService.java:
```diff
@@ -0,0 +1,23 @@
+        this.blockedDomains = Arrays.stream(domains.split(","))
+                .filter(domain -> !domain.isEmpty())
+                .toList();
+    }
```
이건 좀 특이하네요. `,` 로 구분하나요?
- 그렇다면 csv 나 txt 로 하는게 낫지 않나요?

sunsun512 replied:
csv 파일을 업로드 받아 사용하는 것으로 변경하였습니다.

Hyune-c reviewed (changes requested):
step 2-2 맞나요? 2-4 같습니다.

sunsun512 committed:
feat: BlockedDomain 기능
- BlockedDomain을 csvv 파일을 업로드 받아서 하는 것으로 변경
- BlockedDomain을 domain layer단으로 변경

sunsun512 committed:
Merge remote-tracking branch 'origin/blocking_domain' into blocking_domain

# Conflicts:
#	src/main/resources/application-dev.yml

Hyune-c commented on src/main/java/community/whatever/onembackendjava/api/controller/CsvController.java:
```diff
@@ -0,0 +1,29 @@
+        log.info("file >> {}", file);
+        csvReaderService.loadBlockedDomains(file);
+        return ResponseFormatter.ConvertResponse();
+    }
```
```
POST /csv/upload
```

- 전 이게 무슨 기능인지 api endpoint 만 봐서는 전혀 유추되지 않습니다.
- admin 기능에 가까운데 누구나 사용할 수 있는 것 처럼 보입니다.

Hyune-c commented on src/main/java/community/whatever/onembackendjava/api/domain/BlockedDomainInterface.java:
```diff
@@ -0,0 +1,11 @@
+public interface BlockedDomainInterface {
+    void loadBlockedDomains(MultipartFile file);  // CSV 파일에서 차단 목록 로드
+    Set<String> getBlockedDomains();              // 차단된 도메인 목록 반환
+    boolean isBlocked(String url);                // 특정 URL이 차단된 도메인인지 확인
```
`nit` 전 모든 자료형을 wrapper 로 사용합니다.

Hyune-c commented on src/main/java/community/whatever/onembackendjava/api/domain/BlockedDomainInterface.java:
```diff
@@ -0,0 +1,11 @@
+import java.util.Set;
+
+public interface BlockedDomainInterface {
+    void loadBlockedDomains(MultipartFile file);  // CSV 파일에서 차단 목록 로드
```
fyi. javadoc

Hyune-c commented on src/main/java/community/whatever/onembackendjava/api/service/CsvReaderService.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    // URL이 차단된 도메인에 포함되는지 확인
+    public boolean isBlocked(String url) {
+        for (String domain : blockedDomains) {
+            if (url.contains(domain)) {
+                return true;
+            }
+        }
+        return false;
+    }
```
```java
public boolean isBlocked(String url) {
    return blockedDomains.stream().anyMatch(url::contains);
}
```

`nit`

Hyune-c commented on src/main/java/community/whatever/onembackendjava/api/service/CsvReaderService.java:
```diff
@@ -0,0 +1,57 @@
+        try (BufferedReader br = new BufferedReader(
+                new InputStreamReader(file.getInputStream(), StandardCharsets.UTF_8))) {
+
+            String line;
+            while ((line = br.readLine()) != null) {
+                domainSet.add(line.trim());
+            }
```
```java
        blockedDomains.addAll(br.lines()
                                .map(String::trim)
                                .collect(Collectors.toSet()));
```
`nit`

Hyune-c commented on src/main/java/community/whatever/onembackendjava/api/service/CsvReaderService.java:
```diff
@@ -0,0 +1,57 @@
+@Service
+public class CsvReaderService implements BlockedDomainInterface {
+
+    private Set<String> blockedDomains = new HashSet<>();
```
fyi. 일급 컬렉션, holder

sunsun512 committed:
feat: BlockedDomain 기능
- blockedDomains first class
- BlockedDomainInterface javadoc
- blockDomain upload end point 변경
